src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+        // If AD is clear then clear the AD bit.
+
+        // If DO is set then AD is irrelevant. Force AD to be set for
+        // consistency (if DO is set then with respect to the AD flag
+        // the behavior is as if AD is set).
+
+        if key.dnssec_ok {
+            assert!(key.ad);
+        }
+
+        let opt_value = self.cache_lookup_ad(key).await?;
+        if opt_value.is_some() || key.dnssec_ok {
+            return Ok(opt_value);
+        }
+
+        if is_dnssec(key.qtype) {
+            // An explicit request for one of the DNSSEC types but
+            // DO is not set. Force the request to be sent explicitly.
+            return Ok(None);
+        }
+
+        let mut alt_key = key.clone();
+        alt_key.dnssec_ok = true;
+        alt_key.ad = true;
+        let opt_value = self.cache.get(&alt_key).await;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    let msg = remove_dnssec(msg, key.ad)?;
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
+                        )?);
+                    self.cache_insert(key.clone(), value.clone()).await;
+                    return Ok(Some(value));
+                }
+            }
+        }
+        Ok(opt_value)
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// AD flag. The CD, DO, and RD flags are kept unchanged.
+    async fn cache_lookup_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For AD=1 we can only use responses to queries with AD set.
+        // For AD=0, first try with AD=0 and then try with AD=1. If
+        // AD=1 has an answer, clear the AD bit.
+        let opt_value = self.cache.get(key).await;
+        if opt_value.is_some() || key.ad {
+            return Ok(opt_value);
+        }
+        let mut alt_key = key.clone();
+        alt_key.ad = true;
+        let opt_value = self.cache.get(&alt_key).await;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    if !msg.header().ad() {
+                        // AD is clear. Just insert this value.
+                        self.cache_insert(key.clone(), value.clone()).await;
+                        return Ok(Some(value.clone()));
+                    }
+
+                    let msg = clear_ad(msg)?;
+                    let value =
```

Comment:
You are doing this dance of creating a value and then inserting it quite a lot. Maybe it’s worth having a separate method for it?